File: C++/Binary Search Trees/08_preorder_to_bst.cpp
/*
Idea:
using each value from the vector and used the law of BST to insert it.
*/

// C++ Program for the same approach
#include <bits/stdc++.h>
using namespace std;


class Node {
public:
    int data;
    Node* left;
    Node* right;
    Node(int data)
    {
        this->data = data;
        this->left = this->right = NULL;
    }
};

static Node* node;

// This will create the BST
Node* createNode(Node* node, int data)
{
    if (node == NULL)
        node = new Node(data);

    if (node->data > data)
        node->left = createNode(node->left, data);
    if (node->data < data)
        node->right = createNode(node->right, data);

    return node;
}

// A wrapper function of createNode
void create(int data) { node = createNode(node, data); }
// A function to print BST in inorder
void inorderRec(Node* root)
{
    if (root != NULL) {
        inorderRec(root->left);
        cout << root->data << " ";
        inorderRec(root->right);
    }
}

// Driver code
int main()
{
    vector<int> nodeData = { 10, 5, 1, 7, 40, 50 };

    for (int i = 0; i < nodeData.size(); i++) {
        create(nodeData[i]);
    }
    inorderRec(node);
}

// This code is contributed by shinjanpatra

File: Python/Binary Search Trees/08_preorder_to_bst.py
"""
Idea:
using each value from the list and used the law of BST to insert it.
"""


# Construct a BST from given pre-order traversal


class Node:
    data = 0
    left = None
    right = None

    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class CreateBSTFromPreorder:
    node = None

    # This will create the BST

    @staticmethod
    def createNode(node, data):
        if (node == None):
            node = Node(data)
        if (node.data > data):
            node.left = CreateBSTFromPreorder.createNode(node.left, data)
        if (node.data < data):
            node.right = CreateBSTFromPreorder.createNode(node.right, data)
        return node

    # A wrapper function of createNode
    @staticmethod
    def create(data):
        CreateBSTFromPreorder.node = CreateBSTFromPreorder.createNode(
            CreateBSTFromPreorder.node, data)

    # A function to print BST in inorder
    @staticmethod
    def inorderRec(root):
        if (root != None):
            CreateBSTFromPreorder.inorderRec(root.left)
            print(root.data)
            CreateBSTFromPreorder.inorderRec(root.right)

    # Driver Code
    @staticmethod
    def main(args):
        nodeData = [10, 5, 1, 7, 40, 50]
        i = 0
        while (i < len(nodeData)):
            CreateBSTFromPreorder.create(nodeData[i])
            i += 1
        CreateBSTFromPreorder.inorderRec(CreateBSTFromPreorder.node)


if __name__ == "__main__":
    CreateBSTFromPreorder.main([])

